logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb

```diff
@@ -218,16 +218,16 @@ def take_originals_from_queue(queue, size, wait)
           if event.nil?
             # queue poll timed out
             next
-          elsif event == LogStash::SHUTDOWN
+          elsif event.is_a?(LogStash::SignalEvent)
             # We MUST break here. If a batch consumes two SHUTDOWN events
             # then another worker may have its SHUTDOWN 'stolen', thus blocking
             # the pipeline.
-            @shutdown_signal_received = true
-            break
-          elsif event == LogStash::FLUSH
+            @shutdown_signal_received = event.shutdown?
```

Comment:
btw, if this is a concern then we can just place `if event.is_a?(LogStash::Event)` as the first condition